lib/src/steps/result.dart

```diff
@@ -0,0 +1,312 @@
+    String level,
+    String title,
+    String description,
+    SourceSpan loc,
+    double penalty,
+  ) {
+    ArgumentError.checkNotNull(title, 'title');
+    ArgumentError.checkNotNull(description, 'decription');
+    ArgumentError.checkNotNull(penalty, 'penalty');
+    if (title.isEmpty) {
+      throw ArgumentError.value(title, 'title', 'cannot be an empty string');
+    }
+    if (description.isEmpty) {
+      throw ArgumentError.value(
+          description, 'description', 'cannot be an empty string');
+    }
+
+    // We don't want production errors because penalty is out-of-range, instead
+    // we just limit to the range and have an assert that it's in our range.
+    assert(0.0 <= penalty && penalty <= 1.0);
+    penalty = min(max(penalty, 1.0), 0.0);
+
+    _suggestions.add(Suggestion(
+      kind != null ? '$_suggestionCodePrefix.$kind' : _suggestionCodePrefix,
+      level,
+      title,
+      description,
+      file: loc?.sourceUrl?.path,
+      score: 100 - (penalty * 100),
+    ));
+  }
+
+  /// Add [tag] to the result for the current package.
+  ///
+  /// On `pub.dev` package search queries can be filter by tags, such that it is
+  /// possible to limit a search query to packages with a given tag. Or exclude
+  /// packages which has a given tag.
+  ///
+  /// Tags must match a [TagPattern] registered in [tagPatterns], this ensures
+  /// that all tags are documented, and prevents the namespace from overlapping.
+  void tag(String tag) {
+    if (!isValidTag(tag)) {
+      throw ArgumentError.value(
+          tag, 'tag', 'tag is invalid, ensure the pattern is registered');
+    }
+    _tags.add(tag);
+  }
+
+  /// Create a [Result] from a list of [builders].
+  static Result build(Iterable<ResultBuilder> builders) {
+    return Result._(
+      tags: UnmodifiableSetView(builders
+          .map((b) => b._tags)
+          .reduce((s1, s2) => {...s1, ...s2})
+          .toSet()),
+      suggestions: UnmodifiableListView(
+        builders
+            .map((b) => b._suggestions)
+            .reduce((s1, s2) => [...s1, ...s2])
+            .toSet()
+            .toList(),
+      ),
+    );
+  }
+}
+
+/// The [BreakStep] exception can be thrown from an [Step] to abort analysis.
+///
+/// If the [Step] was invoked from [run] the suggestion from the [BreakStep]
+/// exception will be added to the [ResultBuilder]. This is a convinient way to
+/// report a suggestion and return in a single statement, that even works from
+/// an auxiliary method.
+///
+/// Step authors can also catch a [BreakStep] exception from a utility function
+/// and either report it with [report], `rethrow` or ignore it. When caught
+/// step authors can use the [kind] to determine which action to take.
+class BreakStep implements Exception {
+  final String kind;
+  final String _level;
+  final String _title;
+  final String _description;
+  final SourceSpan _loc;
+  final double _penalty;
+
+  BreakStep._({
+    this.kind,
+    String level,
```

Comment:
I imagine that throwing an exception should happen when we find a serious issue, and the `kind`, `level` and `penalty` should be `@required`.